pkg/controllers/nodeoverlay/validation/controller.go

```diff
@@ -0,0 +1,352 @@
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/awslabs/operatorpkg/reasonable"
+	"github.com/samber/lo"
+	"go.uber.org/multierr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	v1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+	"sigs.k8s.io/karpenter/pkg/apis/v1alpha1"
+	"sigs.k8s.io/karpenter/pkg/cloudprovider"
+	"sigs.k8s.io/karpenter/pkg/operator/injection"
+	"sigs.k8s.io/karpenter/pkg/scheduling"
+)
+
+type InstanceTypeUpdate struct {
+	instanceTypeName string
+	nodePoolName     string
+	overlayName      string
+	Offering         map[string]*string
+	Capacity         corev1.ResourceList
+	weight           *int32
+}
+
+type InstanceTypeOverlayStore map[string]*InstanceTypeUpdate
+
+func (s InstanceTypeOverlayStore) FindOverlaidInstanceTypes(nodePool v1.NodePool, nodeOverlay v1alpha1.NodeOverlay, it *cloudprovider.InstanceType) (*InstanceTypeUpdate, bool) {
+	item, found := s[key(nodePool, nodeOverlay, it)]
+	return item, found
+}
+
+func (s InstanceTypeOverlayStore) AddInstanceTypeUpdates(nodePool v1.NodePool, nodeOverlay v1alpha1.NodeOverlay, it *cloudprovider.InstanceType, offerings cloudprovider.Offerings) {
+	// if both are not defined, price will be set to nil
+	price := lo.Ternary(nodeOverlay.Spec.Price == nil, nodeOverlay.Spec.PriceAdjustment, nodeOverlay.Spec.Price)
+	offeringMap := map[string]*string{}
+	if price != nil {
+		for _, of := range offerings {
+			offeringMap[of.Requirements.String()] = price
```

Comment:
I'm wary of using `Requirements.String()` as an index - while it serves the purpose of ensuring we have a unique index per offering, it's a pretty expensive function. Given how often we're going to be iterating over it, it doesn't seem like a great choice. Of course without benchmarking this is just an educated guess, but I think we're going to want a cheaper unique identifier. The simplest option would be to add a new field to offering, ID, which has to be unique within an instance type. Honestly, I don't think that's too unreasonable, it's easy and cheap for the cloudprovider to specify.

Reply:
We have dropped this section, in favor of using an indexed list. Dropping this for now